### PR TITLE
Update parallelism.md

### PR DESCRIPTION
**Pull request by hyunwoongko**
Modifying flexflow part.

### reviewers
@stas00

### CODE REVIEW

File: docs/source/parallelism.md [hunk 1]
```diff
@@ -296,12 +296,27 @@ Paper: ["Beyond Data and Model Parallelism for Deep Neural Networks" by Zhihao J
...
+
+If we perform layer normalization, we compute std first and mean second, finally we can normalize data. At this time, computing std and mean can be parallelize. So if we parallelize them by operator dimension into 2 devices (cuda:0, cuda:1), firstly we copied input data into both devices, and cuda:0 computes std, cuda:1 computes mean at the same time.
+
+* Attribute
+
+We have 10 batches of 512 length. If we parallelize them by attribute dimension into 2 devices, 10 x 512 will be 10 x 2 x 256.
+
+* Parameter
+
+It is similar with tensor model parallelism or niave layer-wise model parallelism.
```
Comment (stas00) [inline docs/source/parallelism.md:319]:
```suggestion
It is similar with tensor model parallelism or naive layer-wise model parallelism.
```

File: docs/source/parallelism.md [hunk 2]
```diff
@@ -296,12 +296,27 @@ Paper: ["Beyond Data and Model Parallelism for Deep Neural Networks" by Zhihao J
...
-2. Operator = part vertical Layer Parallelism, but it can split the layer too - more refined level
-3. Attribute = horizontal Model Parallelism (Megatron-LM style)
-4. Parameter = Sharded model params
+1. Sample = Data Parallelism (sample-wise parallel)
+2. Operator = Parallelize one operation into several tasks
+3. Attribute = Data Parallelism (length-wise parallel)
+4. Parameter = Model Parallelism (it doesn't matter dimension like horizontal or vertical)
 
-and they are working on Pipeline Parallelism. I guess ZeRO-DP is Sample+Parameter in this context.
+### Examples
```
Comment (stas00) [inline docs/source/parallelism.md:304]:
```suggestion
Examples:
```

File: docs/source/parallelism.md [hunk 3]
```diff
@@ -296,12 +296,27 @@ Paper: ["Beyond Data and Model Parallelism for Deep Neural Networks" by Zhihao J
...
+* Sample
+
+We have 10 batches of 512 length. If we parallelize them by sample dimension into 2 devices, 10 x 512 will be 5 x 2 x 512.
+
+* Operator
+
+If we perform layer normalization, we compute std first and mean second, finally we can normalize data. At this time, computing std and mean can be parallelize. So if we parallelize them by operator dimension into 2 devices (cuda:0, cuda:1), firstly we copied input data into both devices, and cuda:0 computes std, cuda:1 computes mean at the same time.
```
Comment (stas00) [inline docs/source/parallelism.md:311]:
```suggestion
If we perform layer normalization, we compute std first and mean second, and then we can normalize data. Operator parallelism allows computing std and mean in parallel. So if we parallelize them by operator dimension into 2 devices (cuda:0, cuda:1), first we copy input data into both devices, and cuda:0 computes std, cuda:1 computes mean at the same time.
```

File: docs/source/parallelism.md [hunk 4]
```diff
@@ -296,12 +296,27 @@ Paper: ["Beyond Data and Model Parallelism for Deep Neural Networks" by Zhihao J
...
+1. Sample = Data Parallelism (sample-wise parallel)
+2. Operator = Parallelize one operation into several tasks
+3. Attribute = Data Parallelism (length-wise parallel)
+4. Parameter = Model Parallelism (it doesn't matter dimension like horizontal or vertical)
 
-and they are working on Pipeline Parallelism. I guess ZeRO-DP is Sample+Parameter in this context.
+### Examples
+* Sample
+
+We have 10 batches of 512 length. If we parallelize them by sample dimension into 2 devices, 10 x 512 will be 5 x 2 x 512.
```
Comment (stas00) [inline docs/source/parallelism.md:307]:
```suggestion
Let's take 10 batches of sequence length 512. If we parallelize them by sample dimension into 2 devices, we get 10 x 512 which becomes be 5 x 2 x 512.
```

File: docs/source/parallelism.md [hunk 5]
```diff
@@ -296,12 +296,27 @@ Paper: ["Beyond Data and Model Parallelism for Deep Neural Networks" by Zhihao J
 
 It performs a sort of 4D Parallelism over Sample-Operator-Attribute-Parameter.
 
-1. Sample = Data Parallelism
-2. Operator = part vertical Layer Parallelism, but it can split the layer too - more refined level
-3. Attribute = horizontal Model Parallelism (Megatron-LM style)
-4. Parameter = Sharded model params
+1. Sample = Data Parallelism (sample-wise parallel)
+2. Operator = Parallelize one operation into several tasks
```
Comment (stas00) [inline docs/source/parallelism.md:300]:
```suggestion
2. Operator = Parallelize a single operation into several sub-operations
```

File: docs/source/parallelism.md [hunk 6]
```diff
@@ -296,12 +296,27 @@ Paper: ["Beyond Data and Model Parallelism for Deep Neural Networks" by Zhihao J
...
+3. Attribute = Data Parallelism (length-wise parallel)
+4. Parameter = Model Parallelism (it doesn't matter dimension like horizontal or vertical)
```
Comment (stas00) [inline docs/source/parallelism.md:302]:
```suggestion
4. Parameter = Model Parallelism (regardless of dimension - horizontal or vertical)
```

**Review by stas00 — APPROVED** [review]
Thank you for updating this summary, @hyunwoongko 

I proposed a few minor wording tweaks.

**Comment by hyunwoongko** [comment]
all is good!

**Pull request merged**